[ADD][16.0] product_print_category

legalsylvain opened the pull request:
### Description

This module is designed to extend product printing features. It allows user to create new print categories of products depending of the data that are on the labels of the products variants.

When a field has changed, the product will be set as 'To print' automatically.

This module requires to install custom modules to have pricetags templates, or alternatively to create pricetag reports templates by UI.

### Configure

* Add yourself to the 'Print Category Manager' group in the 'Pricetags' category.

* Go to 'Sales' > 'Products' > 'Print Categories'

* Find or create new print categories of products
    * Set the fields that will trigger to print again the product label (e.g : name, sale price etc.)
    * Set the qweb view that will be used for this category

![product_print_category_form](https://user-images.githubusercontent.com/3407482/197622734-8e5dd15c-11e2-4e2b-9727-87266217f492.png)


Alternatively, you can create product print categories and qweb in a custom module.

**To affect products to the categories**

* Go to "Sales" > "Products" > "Product variants" (or "Products" if variants are not activated).
* Go to your product in General Information tab
* Set a print category

![product_product_form](https://user-images.githubusercontent.com/3407482/197622683-f8ab9bf0-a486-4e0e-9075-b89994ab621b.png)


**Create Print Category Rules**

* Go to 'Sales' > 'Products' > 'Print Category Rules'

Create various ordered rules depending on product category and company.

This system allow to automatically set / remove print category to a product, when changing its company or its category.

![product_print_category_rule_tree](https://user-images.githubusercontent.com/3407482/213249005-31bd6676-8ec8-47fa-9d65-e31dc49ec8fe.png)




### Usage

* Go to "Sales" > "Products" > "Print Categories"
* Choose between "Print Obsolete Products" or "Print All Products"

![product_print_wizard_form](https://user-images.githubusercontent.com/3407482/197622627-b9346b57-f8c0-4520-afed-380a95535fbc.png)


* Alternatively you can select product in product or variants list view and click on "Action" > "Print Products"

CC :  @remytms I tried to backport this patch https://github.com/grap/grap-odoo-incubator/pull/149/files in V16 but it doesn't seem to work. any idea ?

legalsylvain committed:
[ADD] product_print_category : new module to set print categories on product and allow to print various format of labels

remytms commented:
@legalsylvain I seams to work on my local computer with a fresh 16.0 installed with demo-data.

This is the patch I applied before testing:
```
diff --git a/product_print_category/report/report_pricetag.xml b/product_print_category/report/report_pricetag.xml
index 1be2a8a..37b01f8 100644
--- a/product_print_category/report/report_pricetag.xml
+++ b/product_print_category/report/report_pricetag.xml
@@ -11,7 +11,7 @@ License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
             <t t-foreach="categories_data" t-as="category_data">
                 <div class="page">
                     <t
-                        t-call="#{ category_data['print_category'].qweb_view_id.xml_id }"
+                        t-call="#{ category_data['print_category'].qweb_view_id.key }"
                     />
                 </div>
             </t>
```

This is the PDF report I get: 
[Pricetags.pdf](https://github.com/OCA/product-attribute/files/9870510/Pricetags.pdf)

remytms commented:
@legalsylvain Tell me which error you get when using my patch, if you encounter some. Or a way to reproduce it. :)

legalsylvain committed:
[IMP] replace xml_id by key to make possible to use qweb created via UI, cortesy @remytms

legalsylvain commented:
> @legalsylvain Tell me which error you get when using my patch, if you encounter some. Or a way to reproduce it. :)

That's weird ! I tested again, and all seems to work properly. I added the last commit to backport your improvment.

Thanks ! 

Note : could you review the rest of the PR ?

remytms commented:
For sure I can do a review. :)

remytms commented on product_print_category/README.rst:
```diff
@@ -0,0 +1,91 @@
+Maintainers
+~~~~~~~~~~~
+
+This module is part of the `grap/grap-odoo-incubator <https://github.com/grap/grap-odoo-incubator/tree/12.0/product_print_category>`_ project on GitHub.
```
I think that theses links to https://github.com/grap/grap-odoo-incubator should be replaced by the OCA one, has this module will be in OCA now. :)

legalsylvain replied:
No need. This text is autogenerated by bot when merging. Only fragment should be reviewed

(but done) ;-)

remytms replied:
Thanks for the info. :)

remytms commented on product_print_category/__manifest__.py:
```diff
@@ -0,0 +1,45 @@
+    "maintainers": ["legalsylvain"],
+    "author": "GRAP, "
+    "La Louve, "
+    "Coop IT Easy SCRLfs, "
```
Can you change this to `Coop IT Easy SC` (juridic name reform has occurred in Belgium).

legalsylvain replied:
done.

remytms reviewed (changes requested):
Read all the code and tested it with demo data. Just some naming changes requested.

legalsylvain commented:
@PierrickBrun could you review this one ?

PierrickBrun reviewed (commented):
The product.template view is broken

![Capture d’écran du 2022-12-16 16-31-35](https://user-images.githubusercontent.com/8819682/208133081-2b81a64a-28e4-4ef9-9c21-2a7b330023eb.png)

Apart from that everything looks functional

legalsylvain commented:
@PierrickBrun : I fixed the view issue. could you take a look, and update your review ? 
thanks !

legalsylvain committed:
[IMP] Add onchange feature to make the UI more simple for end user and avoid checkbox to be checked during the save of the product.
[REF] make the module working if a create is done directly on product.template

legalsylvain commented:
Please avoid to merge, I have a final change to do before.

legalsylvain committed:
[IMP] product_print_category : Add product print category rule to have better default print category values.
[REM] remove default_print_category on res.company model

legalsylvain committed:
[REF] product_print_category : do not creeate 'user' print Category Group, because it make the default computation failing on the product.product form view if user is not member of the group. Also to align with the rest of odoo : print_category_id is a regular field that can be user by any user that is member of base.group_user as all the other field
[FIX] use _for_xml_id to avoid error if user is not admin, when using the print wizard
[REF] use expression.AND to simplify code

legalsylvain committed:
[FIX] product_print_category : make wizard working for non admin user

Commit:
[FIX] french trad of print_report_name

rousseldenis commented:
@legalsylvain Still WIP ?

legalsylvain commented:
Hi @rousseldenis That's OK for me !

remytms commented:
@legalsylvain Ping me if you need another review since your last changes. And if I understand correctly, this is not WIP anymore, so you can remove this from the title of this PR. :)

legalsylvain commented:
@rousseldenis could you merge it ?

rousseldenis commented:
@legalsylvain Could you just remove the merge commit ? Thanks

legalsylvain commented:
> @legalsylvain Could you just remove the merge commit ? Thanks

done.

legalsylvain commented:
Hi maintainers : could you merge it ?

carmenbianca commented:
Kind ping @ maintainers to have a look at merging this :)

legalsylvain commented:
/ocabot merge patch

legalsylvain commented:
@OCA/product-maintainers : this one can be merged.

carmenbianca commented:
@OCA/product-maintainers this is still ready to be merged :tada:

bguillot commented:
/ocabot merge patch

dreispt commented:
/ocabot merge nobump

legalsylvain committed:
[FIX] Remove useless apostrophe and fix fr.po file

legalsylvain committed:
[REF] update readme and add index.html file

legalsylvain commented:
It should be OK now. could you retry to merge ?

dreispt commented:
Sure! Shouldn't you be partying?!

/ocabot merge nobump
